src/Aspire.Hosting/Exec/ExecResourceManager.cs

```diff
@@ -0,0 +1,213 @@
+    }
+
+    public async IAsyncEnumerable<CommandOutput> StreamExecResourceLogs([EnumeratorCancellation] CancellationToken cancellationToken)
+    {
+        if (!_execOptions.Enabled)
+        {
+            yield break;
+        }
+
+        string type = "waiting";
+
+        yield return new CommandOutput
+        {
+            Text = $"Waiting for resources to be initialized...",
+            Type = type
+        };
+
+        // wait until AppHost eventing fires ConfigureExecResource()
+        // and execResource is initialized
+        IResource? execResource = null;
+        Exception? execResourceInitializationException = null;
+        try
+        {
+            execResource = await _execResourceInitialized.Task.WaitAsync(cancellationToken).ConfigureAwait(false);
+        }
+        catch (OperationCanceledException)
+        {
+            _logger.LogInformation("Cancelled before exec resource was initialized.");
+            yield break;
+        }
+        catch (Exception ex)
+        {
+            _logger.LogError(execResourceInitializationException, "Exec resource initialization failed.");
+            execResourceInitializationException = ex;
+        }
+
+        if (execResourceInitializationException is not null)
+        {
+            yield return new CommandOutput
+            {
+                Text = execResourceInitializationException.Message,
+                IsErrorMessage = true,
+                Type = type
+            };
+            yield break;
+        }
+
+        // dcp annotation is populated by other handler of BeforeStartEvent
+        var dcpExecResourceName = execResource!.GetResolvedResourceName();
+
+        yield return new CommandOutput
+        {
+            Text = $"Aspire exec starting...",
+            Type = type
+        };
+
+        // in the background wait for the exec resource to be running to change log type
+        _ = Task.Run(async () =>
+        {
+            await _resourceNotificationService.WaitForResourceAsync(execResource!.Name, targetState: KnownResourceStates.Running, cancellationToken).ConfigureAwait(false);
+            type = "running";
+        }, cancellationToken);
+
+        // in the background wait for the exec resource to reach terminal state. Once done we can complete logging
+        _ = Task.Run(async () =>
+        {
+            await _resourceNotificationService.WaitForResourceAsync(execResource!.Name, targetStates: KnownResourceStates.TerminalStates, cancellationToken).ConfigureAwait(false);
+
+            // hack: https://github.com/dotnet/aspire/issues/10245
+            // workarounds the race-condition between streaming all logs from the resource, and resource completion
+            await Task.Delay(1000, CancellationToken.None).ConfigureAwait(false);
+
+            _resourceLoggerService.Complete(dcpExecResourceName); // complete stops the `WatchAsync` async-foreach below
+        }, cancellationToken);
```

Comment:
@karolz-ms do you have any recommendations here?

Reply:
I had a conversation with @DeagleGross about this, and I think it is fine to take this implementation as is is for now, but we should use the issue that Dmitry has opened to dig in and understand why `WatchAsync` does not complete for resources that reach the final state (which is the case here) because I think it should, and I explained why in the related issue.